Initial Client external transport API (#51)
eProsima/Micro-XRCE-DDS-docs: docs/transport.rst

    size_t uxr_write_udp_data_platform(uxrUDPPlatform* platform, const uint8_t* buf, size_t len, uint8_t* errcode);
    size_t uxr_read_udp_data_platform(uxrUDPPlatform* platform, uint8_t* buf, size_t len, int timeout, uint8_t* errcode);

Stream Framing Protocol
-----------------------

*eProsima Micro XRCE-DDS* has a **Stream Framing Protocol** with the following features:

* **HDLC Framing**: each frame begins with a ``begin_frame`` octet ``(0x7E)``, and the rest of the frame is byte stuffing using the ``space`` octet ``(0x7D)`` following by the original octet exclusive-or with ``0x20``.
  For example, if the frame contains the octet `0x7E` it is encoded as `0x7D, 0x5E`; and the same for the octet `0x7E` which is encoded as `0x7D, 0x5D`.
* **CRC Calculation**: frames end with the CRC-16 for detecting frame corruption.
  The CRC-16 is computed using the polynomial ``x^16 + x^12 + x^5 + 1`` after the frame stuffing for each octet of the frame and including the ``begin_frame``, as it is described in `RFC 1662 <https://tools.ietf.org/html/rfc1662>`_ (see sec. C.2).

> * **Routing header**: the Stream Framing Protocol provides ``source`` and ``remote`` address in the framing, which could be used for implement a routing protocol.

Reviewer: ```suggestion
* **Routing header**: the Stream Framing Protocol provides ``source`` and ``remote`` addresses in the framing, which can be used to implement a routing protocol.
```

    1. A publisher application calls the *Client* library to send a given topic.
    2. The *Client* library serializes the topic inside an XRCE message using *Micro CDR*.
       As a result, the XRCE message with the topic is stored in an **Output Stream Buffer**.
    3. The *Client* library calls the Stream Framing Protocol to send the serialized message.

>     4. The Stream Transport frames the message, that is, adds the header, payload, and CRC of the frame, taking into account the stuffing.

Reviewer: ```suggestion
    4. The Stream Transport frames the message, that is, it adds the header, the payload, and CRC of the frame, taking into account the stuffing.
```

       This step takes place in an auxiliary buffer called **Framing Buffer**.
    5. Each time the Framing Buffer is full, the data is flushed into the **Device Buffer**, calling the writing system function.

.. image:: images/serial_transport_sending.svg

This approach has some advantages which should be pointed out:

    1. The HDLD framing and the CRC control provide **integrity** and **security** to the Stream Framing.

>     2. The framing technique allows to **reducing memory usage**.

Reviewer: ```suggestion
    2. The framing technique allows to **reduce memory usage**.
```

>        It is because the Framing Buffer size (42 bytes) bounds the Device Buffer size.

Reviewer: ```suggestion
       The reason is that the Framing Buffer size (42 bytes) bounds the Device Buffer size.
```

>     3. The framing technique also allows sending **large data** over stream oriented transports.

Reviewer: ```suggestion
    3. The framing technique also allows sending **large data** over stream-oriented transports.
```

>        It is because the message size is not bounded by the Device Buffer size, since the message is fragmented and stuffing during the framing stage.

Reviewer: ```suggestion
       The reason is that the message size is not bounded by the Device Buffer size, since the message is fragmented and has undergone byte stuffing during the framing stage.
```

The figure above shows the overhead added by Serial Transport.
In the best case, it is **only 19 bytes**, but it should be noted that, in this example, the message stuffing has been neglected.

Custom Transport API
--------------------

> XRCE-DDS provides an user API that allows interfacing with the lowest level transport layer at runtime. This way an user is enabled to implement its own transports based on one of the two communication approaches: stream-oriented or packet-oriented.

Reviewer: ```suggestion
*eProsima Micro XRCE-DDS* provides a user API that allows interfacing with the lowest level transport layer at runtime. In this way, a user is enabled to implement its own transports based on one of the two communication approaches: stream-oriented or packet-oriented.
```

Client custom transport API
^^^^^^^^^^^^^^^^^^^^^^^^^^^

> By means of using the following functions an user can set four callbacks that will be in charge of opening and closing the transport, and writing and reading from it. This custom transport API is enabled by setting the CMake argument ``UCLIENT_PROFILE_CUSTOM_TRANSPORT=<bool>`` to true. In the case that stream-oriented transport is used ``UCLIENT_PROFILE_STREAM_FRAMING=<bool>`` should also be enabled.

Reviewer: ```suggestion
By means of the following functions, a user can set four callbacks which will be in charge of opening and closing the transport, and writing and reading from it. This custom transport API is enabled by setting the CMake argument ``UCLIENT_PROFILE_CUSTOM_TRANSPORT=<bool>`` to true. In the case that stream-oriented transport is used ``UCLIENT_PROFILE_STREAM_FRAMING=<bool>`` should also be enabled.
```

UXRDLLAPI void uxr_set_custom_transport_callbacks(uxrCustomTransport* transport,
                                                  bool framing,
                                                  open_custom_func open,
                                                  close_custom_func close,
                                                  write_custom_func write,
                                                  read_custom_func read);
------

.. code-block:: c

    void uxr_set_custom_transport_callbacks(uxrCustomTransport* transport, bool framing, open_custom_func open, close_custom_func close, write_custom_func write, read_custom_func read);

Assigns the callback for custom transport.

> 

Reviewer: isn't the description of :transport: missing here?

Author: Solved

:framing: Enables or disables Stream Framing Protocol for a custom transport.
:open: Callback for opening a custom transport.
:close: Callback for closing a custom transport.
:write: Callback for writing to a custom transport.
:read: Callback for reading from a custom transport.

The function signatures for the above callbacks are:

.. code-block:: c

    typedef bool (*open_custom_func)(struct uxrCustomTransport*);

Where ``struct uxrCustomTransport*`` will have the ``args`` passed through ``bool uxr_init_custom_transport(uxrCustomTransport* transport, void * args);``

.. code-block:: c

    typedef bool (*close_custom_func)(struct uxrCustomTransport*);

Where ``struct uxrCustomTransport*`` will have the ``args`` passed through ``bool uxr_init_custom_transport(uxrCustomTransport* transport, void * args);``

.. code-block:: c

    typedef size_t (*write_custom_func)(struct uxrCustomTransport*, const uint8_t*, size_t, uint8_t*);

> Where ``struct uxrCustomTransport*`` refers to the opened transport structure,  the first ``const uint8_t*`` is the buffer to be sent, following ``size_t`` is the length of the buffer and the last ``uint8_t*`` is an error code that should be set when the write process has any problem. This function should return the number of bytes sent successfully.

Reviewer: ```suggestion
Where ``struct uxrCustomTransport*`` refers to the opened transport structure,  the first ``const uint8_t*`` is the buffer to be sent, ``size_t`` is the length of the buffer, and the last ``uint8_t*`` is an error code that should be set when the write process has any problem. This function should return the number of bytes sent successfully.
```
